```
 ... fail_compilation/fail_arrayop3c.d -verrors=0 -o-  -fPIC ()
+ check_clean_git
+ git checkout test/compilable/issue17167.sh
+ rm -rf _generated
+ rm -f install.sh
+ rm -f test/compilable/vcg-ast.d.cg
+ make -f posix.mak check-clean-git
ERROR: Found the following residual temporary files.
ERROR: Temporary files should be stored in `test_results` or explicitly removed.
?? test/compilable/vcg-ast-arraylength.d.cg
posix.mak:69: recipe for target 'check-clean-git' failed
make: *** [check-clean-git] Error 1
```

--- a/.circleci/run.sh
+++ b/.circleci/run.sh
@@ -150,6 +150,7 @@ check_clean_git()
     rm -f install.sh
     # auto-removal of these files doesn't work on CirleCi
     rm -f test/compilable/vcg-ast.d.cg
+    rm -f test/compilable/vcg-ast-arraylength.d.cg
     # Ensure that there are no untracked changes
     make -f posix.mak check-clean-git
 }

--- a/src/dmd/expressionsem.d
+++ b/src/dmd/expressionsem.d
@@ -9137,6 +9137,16 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (!verifyHookExist(exp.loc, *sc, Id._d_arraysetlengthTImpl, "resizing arrays"))
                 return setError();
 
+            // use slice expression when arr.length = 0 to avoid runtime call
+            if(exp.e2.isConst() && exp.e2.toUInteger() == 0)
+            {
+                IntervalExp ie = new IntervalExp(ale.loc, exp.e2, exp.e2);
+                Expression se = new SliceExp(ale.loc, ale.e1, ie);
+                Expression as = new AssignExp(ale.loc, ale.e1, se);
+                auto res = as.expressionSemantic(sc);
+                return setResult(res);
+            }
+
             // Lower to object._d_arraysetlengthTImpl!(typeof(e1))._d_arraysetlengthT{,Trace}(e1, e2)
             Expression id = new IdentifierExp(ale.loc, Id.empty);
             id = new DotIdExp(ale.loc, id, Id.object);

--- a/test/compilable/extra-files/vcg-ast-arraylength-postscript.sh
+++ b/test/compilable/extra-files/vcg-ast-arraylength-postscript.sh
@@ -1,0 +1,12 @@
+#!/usr/bin/env bash
+
+source tools/common_funcs.sh
+
+# Test if there's call to the runtime for .length = 100
+grep "_d_arraysetlengthT(arr, 100.*)" "${TEST_DIR}/${TEST_NAME}.d.cg"
+# Make sure there's no call to the runtime for .length = 0
+! grep "_d_arraysetlengthT(arr, 0.*)" "${TEST_DIR}/${TEST_NAME}.d.cg"
+# Test if a slice expr is applied for the above case
+grep "arr = arr\[0..0\]" "${TEST_DIR}/${TEST_NAME}.d.cg"
+
+rm_retry "${TEST_DIR}/${TEST_NAME}.d.cg"

--- a/test/compilable/vcg-ast-arraylength.d
+++ b/test/compilable/vcg-ast-arraylength.d
@@ -1,0 +1,15 @@
+module vcg_ast_arraylength;
+// REQUIRED_ARGS: -vcg-ast -o-
+// PERMUTE_ARGS:
+// POST_SCRIPT: compilable/extra-files/vcg-ast-arraylength-postscript.sh
+
+void main()
+{
+	int[] arr = [1, 2, 3];
+
+	// may use runtime call
+	arr.length = 100;
+
+	// should convert to arr = arr[0 .. 0];
+	arr.length = 0;
+}

